_posts/2019/2019-12-01-ensure-you-correctly-build-caching-keys.md

```diff
@@ -0,0 +1,112 @@
+---
+title: "Ensure you correctly build your caching keys"
+description: "Why won't this model update? Oh."
+layout: article
+category: ruby
+image:
+  base: '2019/ensure-you-correctly-build-caching-keys'
+  alt: 'Keys'
+  credit:  Chunlea Ju
+  source: "https://unsplash.com/photos/8fs1X0JFgFE"
+
+---
+
+Caching is a hugely powerful tool in maintaining the performance of often requested pages and partials in your application.
+
+It can also cause confusing behaviour (mostly “stuff not updating on the page”) when you test your app in a production environment.
+
+Rails has great built-in support for many types of caching, in particular the “view fragment caching”. This stores the resulting text  of parts your views in very fast storage, say Redis or Memcache, saving your application building the views every time a page is rendered.
+
+The framework includes an elegant way of using the cache based on using the model `id`, it’s `updated_at` timestamp and an automatically generated digest of the specific view template. You can find more details in the [Rails guide on caching](https://guides.rubyonrails.org/caching_with_rails.html#fragment-caching).
+
+
+## Instead of...
+
+...just using a single modal in your key, for a complex view
+
+```ruby
+class Event < ApplicationRecord
+  has_many :attendees
+end
+
+class Attendee < ApplicationRecord
+  belongs_to :event
+  has_many :prerequistes
+end
+
+class Prerequistes < ApplicationRecord
+  belongs_to :attendee
+
+  scope :done, -> { where.not(completed_at: nil) }
+end
+```
+
+
+### `app/views/events/show.html.erb`
+
+```erb
+<% @event.attendees.each do |attendee| %>
+  <% cache [attendee] do %>
+    <p>
+      <%= attendee.name %>, <%= @event.name %><br />
+      <small>
+        <%= attendee.prerequistes.done.count %> done
+      </small>
+    </p>
+  <% end %>
+<% end %>
+```
+
+
+## Use
+
+...multiple relevant models, including the parent object.
+
+```ruby
+class Event < ApplicationRecord
+  has_many :attendees
+end
+
+class Attendee < ApplicationRecord
+  belongs_to :event, touch: true
+  has_many :prerequistes
+end
+
+class Prerequistes < ApplicationRecord
+  belongs_to :attendee, touch: true
+  scope :done, -> { where.not(completed_at: nil) }
+end
+```
+
+### `app/views/events/show.html.erb`
+
+```erb
+<% @event.attendees.each do |attendee| %>
+  <% cache [@event, attendee] do %>
+    <p>
+      <%= attendee.name %>, <%= @event.name %><br />
+      <small>
+        <%= attendee.prerequistes.done.count %> done
+      </small>
+    </p>
+  <% end %>
+<% end %>
+```
+
+
+## Why?
+
+This is a ‘powerful tools enabling subtle bugs’ problem.
+
+When we use the data from any “parent” model in a view-cached partial, you need to include that model in the cache key, or the view will not update with changes to that parent.
```

Comment:
Is it worth referring to 'parent' in this paragraph as this whole paragraph doesn't refer to any 'children'. It only refers to the model in question. You're basically trying to say: you need to reference any models you want updated in the cache key, right?